src/components/toaster/toaster-context.tsx

```diff
@@ -0,0 +1,63 @@
+import { createContext, useContext, useState } from 'react';
+
+const TOAST_REMOVE_DELAY = 3000;
+type ToastType = 'success' | 'error' | 'warning' | 'info';
+
+interface ToastData {
+    type: ToastType;
+    message: string;
+    isOpen: boolean;
+    timeoutId: number;
+}
+
+export function ToasterContextProvider({ children }: React.PropsWithChildren) {
+    const [toastData, setToastData] = useState<ToastData | null>(null);
+
+    const toast = ({
+        type,
+        message,
+        duration,
+    }: {
+        type: ToastType;
+        message: string;
+        duration?: number;
+    }) => {
+        setToastData({
+            type,
+            message,
+            timeoutId: window.setTimeout(closeToast, duration ?? TOAST_REMOVE_DELAY),
+            isOpen: true,
+        });
+    };
+
+    const closeToast = () => {
+        setToastData((toast) => {
+            if (toast && toast.timeoutId) window.clearTimeout(toast.timeoutId);
+            return null;
+        });
+    };
+
+    const onOpenChange = () => (isOpen: boolean) => {
+        setToastData((prevToastData) => (prevToastData ? { ...prevToastData, isOpen } : null));
+    };
+
+    const contextValue = { toastData, toast, closeToast, onOpenChange };
+    return <ToasterContext.Provider value={contextValue}>{children}</ToasterContext.Provider>;
+}
+
+interface ToasterContextType {
```

Comment:
this context type is a bit wierd.
there is `closeToast` but no `showToast`. Also does the `toastData` mean that there could be only one toast opened at a time?

Reply:
> only one toast opened at a time

yes.